sydent/http/servlets/replication.py

```diff
@@ -40,67 +69,139 @@ def render_POST(self, request):
+            request.finish()
+            return
 
         try:
             inJson = json.load(request.content)
         except ValueError:
             logger.warn("Peer %s made push connection with malformed JSON", peer.servername)
-            return {'errcode': 'M_BAD_JSON', 'error': 'Malformed JSON'}
-
-        if 'sgAssocs' not in inJson:
-            logger.warn("Peer %s made push connection with no 'sgAssocs' key in JSON", peer.servername)
-            return {'errcode': 'M_BAD_JSON', 'error': 'No "sgAssocs" key in JSON'}
-
-        failedIds = []
-
-        globalAssocsStore = GlobalAssociationStore(self.sydent)
-
-        for originId,sgAssoc in inJson['sgAssocs'].items():
-            try:
-                peer.verifySignedAssociation(sgAssoc)
-                logger.debug("Signed association from %s with origin ID %s verified", peer.servername, originId)
+            request.setResponseCode(400)
+            request.write(json.dumps({'errcode': 'M_BAD_JSON', 'error': 'Malformed JSON'}))
+            request.finish()
+            return
 
-                # Don't bother adding if one has already failed: we add all of them or none so we're only going to
-                # roll back the transaction anyway (but we continue to try & verify the rest so we can give a
-                # complete list of the ones that don't verify)
-                if len(failedIds) > 0:
-                    continue
+        # Ensure there is data we are able to process
+        if 'sg_assocs' not in inJson and 'invite_tokens' not in inJson and 'ephemeral_public_keys' not in inJson:
+            logger.warn("Peer %s made push connection with no 'sg_assocs', 'invite_tokens' or 'ephemeral_public_keys' keys in JSON", peer.servername)
+            request.setResponseCode(400)
+            request.write(json.dumps({'errcode': 'M_BAD_JSON', 'error': 'No "sg_assocs", "invite_tokens" or "ephemeral_public_keys" key in JSON'}))
+            request.finish()
+            return
+
+        # Check for any signed associations
+        if 'sg_assocs' in inJson and len(inJson['sg_assocs']) > 0:
+            if len(inJson['sg_assocs']) > MAX_SG_ASSOCS_LIMIT:
+                request.setResponseCode(400)
+                request.write(json.dumps({'errcode': 'M_BAD_JSON', 'error': '"sg_assocs" has more than %d keys' % MAX_SG_ASSOCS_LIMIT}))
+                request.finish()
+                return
+
+            globalAssocsStore = GlobalAssociationStore(self.sydent)
+
+            # Check that this message is signed by one of our trusted associated peers
+            for originId, sgAssoc in inJson['sg_assocs'].items():
+                try:
+                    yield peer.verifySignedAssociation(sgAssoc)
+                except (NoSignaturesException, NoMatchingSignatureException, RemotePeerError, SignatureVerifyException):
+                    self.sydent.db.rollback()
+                    logger.warn("Failed to verify signed association from %s", peer.servername)
```

Comment:
we should probably log what the failure was. Also should we not record which association it was that failed?